File: project/SCM_Teleop/cpp_collect/collect_vehicle_data.cpp
// =============================================================================
// Full Vehicle SCM Data Collection for NN Tire Model Training
// =============================================================================
//
// Unlike the tire test rig, this collects data from actual vehicle simulation:
// - Real tire loads from vehicle weight and load transfer
// - Natural slip angles from cornering maneuvers  
// - Forces that match what the estimator sees
//
// This should produce more accurate training data for the terrain estimator.
// =============================================================================

#include <cstdio>
#include <cmath>
#include <vector>
#include <random>
#include <chrono>
#include <fstream>
#include <iomanip>
#include <sstream>
#include <algorithm>
#include <atomic>
#include <mutex>
#include <thread>

#ifdef __linux__
#include <malloc.h>
#endif

#include "chrono/core/ChRealtimeStep.h"
#include "chrono/physics/ChSystemSMC.h"
#include "chrono/solver/ChIterativeSolverLS.h"

#include "chrono_vehicle/ChConfigVehicle.h"
#include "chrono_vehicle/ChVehicleModelData.h"
#include "chrono_vehicle/terrain/SCMTerrain.h"

#include "chrono_models/vehicle/hmmwv/HMMWV.h"

#include "chrono_thirdparty/cxxopts/ChCLI.h"

#ifdef CHRONO_OPENMP
#include <omp.h>
#endif

using namespace chrono;
using namespace chrono::vehicle;
using namespace chrono::vehicle::hmmwv;

// =============================================================================
// Terrain parameter ranges (matching cpp_collect)
// =============================================================================
struct TerrainRanges {
    double bekker_Kphi_min = 0.5e6;
    double bekker_Kphi_max = 4.0e6;
    double bekker_Kc_min = 0;
    double bekker_Kc_max = 20000;
    double bekker_n_min = 0.3;
    double bekker_n_max = 1.5;
    double mohr_cohesion_min = 0;
    double mohr_cohesion_max = 10000;
    double mohr_friction_min = 10;   // degrees
    double mohr_friction_max = 45;   // degrees
    double janosi_shear_min = 0.005;
    double janosi_shear_max = 0.05;
};

// =============================================================================
// Sample parameters for a single run
// =============================================================================
struct TerrainParams {
    double bekker_Kphi;
    double bekker_Kc;
    double bekker_n;
    double mohr_cohesion;
    double mohr_friction;  // degrees
    double janosi_shear;
};

// =============================================================================
// Single tire measurement
// =============================================================================
struct TireMeasurement {
    double time;
    int axle;           // 0 = front, 1 = rear
    int side;           // 0 = left, 1 = right
    double slip_angle;  // radians
    double slip_ratio;
    double camber;      // radians
    double velocity;    // forward velocity m/s
    double Fz;          // vertical force (positive = compression)
    double Fx;          // longitudinal force
    double Fy;          // lateral force
    // Terrain params for this sample
    double bekker_Kphi;
    double bekker_Kc;
    double bekker_n;
    double mohr_cohesion;
    double mohr_friction;  // degrees
    double janosi_shear;
};

// =============================================================================
// Generate random terrain parameters
// =============================================================================
TerrainParams GenerateRandomTerrain(const TerrainRanges& ranges, std::mt19937& rng) {
    std::uniform_real_distribution<double> dist(0.0, 1.0);
    
    TerrainParams params;
    params.bekker_Kphi = ranges.bekker_Kphi_min + dist(rng) * (ranges.bekker_Kphi_max - ranges.bekker_Kphi_min);
    params.bekker_Kc = ranges.bekker_Kc_min + dist(rng) * (ranges.bekker_Kc_max - ranges.bekker_Kc_min);
    params.bekker_n = ranges.bekker_n_min + dist(rng) * (ranges.bekker_n_max - ranges.bekker_n_min);
    params.mohr_cohesion = ranges.mohr_cohesion_min + dist(rng) * (ranges.mohr_cohesion_max - ranges.mohr_cohesion_min);
    params.mohr_friction = ranges.mohr_friction_min + dist(rng) * (ranges.mohr_friction_max - ranges.mohr_friction_min);
    params.janosi_shear = ranges.janosi_shear_min + dist(rng) * (ranges.janosi_shear_max - ranges.janosi_shear_min);
    
    return params;
}

// =============================================================================
// Run a single vehicle simulation and collect tire data
// =============================================================================
std::vector<TireMeasurement> RunVehicleSimulation(
    const TerrainParams& terrain,
    int sample_idx,
    double sim_duration = 15.0,
    double step_size = 3e-3,
    double mesh_spacing = 0.1,
    bool verbose = false
) {
    std::vector<TireMeasurement> measurements;
    
    // Create system
    auto sys = std::make_unique<ChSystemSMC>();
    sys->SetCollisionSystemType(ChCollisionSystem::Type::BULLET);
    sys->SetGravitationalAcceleration(ChVector3d(0, 0, -9.81));
    sys->SetSolverType(ChSolver::Type::BARZILAIBORWEIN);
    sys->SetTimestepperType(ChTimestepper::Type::EULER_IMPLICIT_LINEARIZED);
    sys->GetSolver()->AsIterative()->SetMaxIterations(100);
    
    // Create HMMWV vehicle
    HMMWV_Full hmmwv(sys.get());
    hmmwv.SetCollisionSystemType(ChCollisionSystem::Type::BULLET);
    hmmwv.SetChassisFixed(false);
    hmmwv.SetInitPosition(ChCoordsys<>(ChVector3d(0, 0, 0.6), ChQuaterniond(1, 0, 0, 0)));
    hmmwv.SetEngineType(EngineModelType::SHAFTS);
    hmmwv.SetTransmissionType(TransmissionModelType::AUTOMATIC_SHAFTS);
    hmmwv.SetDriveType(DrivelineTypeWV::AWD);
    hmmwv.SetTireType(TireModelType::RIGID);  // RIGID for SCM interaction
    hmmwv.Initialize();
    
    hmmwv.SetChassisVisualizationType(VisualizationType::NONE);
    hmmwv.SetSuspensionVisualizationType(VisualizationType::NONE);
    hmmwv.SetSteeringVisualizationType(VisualizationType::NONE);
    hmmwv.SetWheelVisualizationType(VisualizationType::NONE);
    hmmwv.SetTireVisualizationType(VisualizationType::NONE);
    
    // Create SCM terrain
    SCMTerrain scm_terrain(sys.get());
    scm_terrain.SetSoilParameters(
        terrain.bekker_Kphi,
        terrain.bekker_Kc,
        terrain.bekker_n,
        terrain.mohr_cohesion,
        terrain.mohr_friction * CH_DEG_TO_RAD,  // Convert to radians
        terrain.janosi_shear,
        2e8,   // elastic stiffness
        3e4    // damping
    );
    
    // Add moving patch for efficiency
    scm_terrain.AddMovingPatch(
        hmmwv.GetChassisBody(),
        ChVector3d(0, 0, 0),
        ChVector3d(6, 3, 1)
    );
    
    scm_terrain.SetPlotType(SCMTerrain::PLOT_NONE, 0, 0);
    scm_terrain.Initialize(100.0, 30.0, mesh_spacing);
    
    // Vehicle parameters
    const double Lf = 1.689;  // front axle to CG
    const double Lr = 1.689;  // rear axle to CG
    
    // Simulation loop
    double t = 0.0;
    int sample_count = 0;
    const int sample_interval = 100;  // Every 100 steps = 0.3s
    
    while (t < sim_duration) {
        // Generate driving inputs: acceleration then cornering
        double steer = 0.0;
        double throttle = 0.0;
        
        if (t < 2.0) {
            // Accelerate straight
            steer = 0.0;
            throttle = 0.5;
        } else if (t < 5.0) {
            // Left turn ramp up
            steer = 0.3 * (t - 2.0) / 3.0;
            throttle = 0.35;
        } else if (t < 8.0) {
            // Hold left turn
            steer = 0.3;
            throttle = 0.35;
        } else if (t < 11.0) {
            // Transition to right
            steer = 0.3 - 0.6 * (t - 8.0) / 3.0;
            throttle = 0.35;
        } else {
            // Hold right turn
            steer = -0.3;
            throttle = 0.35;
        }
        
        // Apply inputs
        DriverInputs inputs;
        inputs.m_steering = steer;
        inputs.m_throttle = throttle;
        inputs.m_braking = 0.0;
        
        // Synchronize and advance
        scm_terrain.Synchronize(t);
        hmmwv.Synchronize(t, inputs, scm_terrain);
        scm_terrain.Advance(step_size);
        hmmwv.Advance(step_size);
        
        // Sample tire data periodically after initial settling
        if (t > 3.0 && (sample_count % sample_interval) == 0) {
            // Get chassis state
            auto chassis = hmmwv.GetChassisBody();
            auto vel_global = chassis->GetPosDt();
            auto rot = chassis->GetRot();
            auto rot_inv = rot.GetInverse();
            auto vel_body = rot_inv.Rotate(vel_global);
            auto omega_vec = chassis->GetAngVelLocal();
            
            double u = vel_body.x();  // forward velocity
            double v = vel_body.y();  // lateral velocity
            double omega = omega_vec.z();  // yaw rate
            double delta = steer * 0.5;  // steering angle (approximate ratio)
            
            // Skip if vehicle too slow
            if (u < 2.0) {
                sample_count++;
                t += step_size;
                continue;
            }
            
            // Get tire forces for all 4 wheels
            for (int axle = 0; axle < 2; axle++) {
                for (int side = 0; side < 2; side++) {
                    auto side_enum = (side == 0) ? VehicleSide::LEFT : VehicleSide::RIGHT;
                    auto tire = hmmwv.GetVehicle().GetTire(axle, side_enum);
                    
                    // Get tire force in global frame, then transform
                    auto tire_force = tire->ReportTireForce(&scm_terrain);
                    auto force_local = rot_inv.Rotate(tire_force.force);
                    
                    // Compute slip angle for this wheel
                    double L = (axle == 0) ? Lf : -Lr;  // + for front, - for rear
                    double wheel_delta = (axle == 0) ? delta : 0.0;  // only front steers
                    
                    // Slip angle: alpha = atan2(v + L*omega, u) - delta
                    double eps = 0.5;
                    double u_safe = std::max(u, eps);
                    double alpha = std::atan2(v + L * omega, u_safe) - wheel_delta;
                    
                    // Store measurement
                    TireMeasurement m;
                    m.time = t;
                    m.axle = axle;
                    m.side = side;
                    m.slip_angle = alpha;
                    m.slip_ratio = 0.0;  // Not computing this accurately for now
                    m.camber = 0.0;
                    m.velocity = u;
                    m.Fz = force_local.z();  // Positive = compression
                    m.Fx = force_local.x();
                    m.Fy = force_local.y();
                    m.bekker_Kphi = terrain.bekker_Kphi;
                    m.bekker_Kc = terrain.bekker_Kc;
                    m.bekker_n = terrain.bekker_n;
                    m.mohr_cohesion = terrain.mohr_cohesion;
                    m.mohr_friction = terrain.mohr_friction;
                    m.janosi_shear = terrain.janosi_shear;
                    
                    // Only keep samples with reasonable forces
                    if (m.Fz > 1000 && std::abs(m.Fy) < 20000 && std::abs(alpha) > 0.01) {
                        measurements.push_back(m);
                    }
                }
            }
        }
        
        sample_count++;
        t += step_size;
    }
    
    // Cleanup
    sys->Clear();
    sys.reset();
    
#ifdef __linux__
    malloc_trim(0);
#endif
    
    if (verbose) {
        std::cout << "Sample " << sample_idx << ": collected " << measurements.size() 
                  << " tire measurements (n=" << terrain.bekker_n 
                  << ", phi=" << terrain.mohr_friction << "°)" << std::endl;
    }
    
    return measurements;
}

// =============================================================================
// Write measurements to CSV
// =============================================================================
void WriteMeasurementsToCSV(std::ofstream& csv_file, std::mutex& csv_mutex,
                            const std::vector<TireMeasurement>& measurements) {
    std::ostringstream buffer;
    buffer << std::fixed << std::setprecision(6);
    
    for (const auto& m : measurements) {
        buffer << m.Fz << ","                 // vertical_load
               << m.slip_angle << ","         // slip_angle (radians)
               << m.slip_ratio << ","         // longitudinal_slip
               << m.camber << ","             // camber_angle
               << m.velocity << ","           // velocity
               << m.bekker_Kphi << ","
               << m.bekker_Kc << ","
               << m.bekker_n << ","
               << m.mohr_cohesion << ","
               << m.mohr_friction << ","      // degrees
               << m.janosi_shear << ","
               << 0.1 << ","                  // mesh_spacing (fixed)
               << m.Fz << ","                 // Fz (same as vertical_load for vehicle)
               << m.Fx << ","
               << m.Fy << "\n";
    }
    
    std::lock_guard<std::mutex> lock(csv_mutex);
    csv_file << buffer.str();
    csv_file.flush();
}

// =============================================================================
// Main function
// =============================================================================
int main(int argc, char* argv[]) {
    // Parse command line
    ChCLI cli(argv[0]);
    cli.AddOption<int>("", "n", "Number of vehicle simulations to run", "100");
    cli.AddOption<std::string>("", "o", "Output CSV file", "vehicle_tire_data.csv");
    cli.AddOption<int>("", "j", "Number of parallel threads", "4");
    cli.AddOption<double>("", "t", "Simulation duration per run (seconds)", "15");
    cli.AddOption<double>("", "mesh", "SCM mesh spacing", "0.1");
    cli.AddOption<bool>("", "v", "Verbose output", "true");
    cli.AddOption<int>("", "seed", "Random seed", "12345");
    
    if (!cli.Parse(argc, argv, true)) {
        cli.Help();
        return 1;
    }
    
    int num_sims = cli.GetAsType<int>("n");
    std::string output_file = cli.GetAsType<std::string>("o");
    int num_threads = cli.GetAsType<int>("j");
    double sim_duration = cli.GetAsType<double>("t");
    double mesh_spacing = cli.GetAsType<double>("mesh");
    bool verbose = cli.GetAsType<bool>("v");
    int seed = cli.GetAsType<int>("seed");
    
    std::cout << "==================================================================" << std::endl;
    std::cout << "Full Vehicle SCM Data Collection" << std::endl;
    std::cout << "==================================================================" << std::endl;
    std::cout << "Simulations: " << num_sims << std::endl;
    std::cout << "Threads: " << num_threads << std::endl;
    std::cout << "Duration each: " << sim_duration << "s" << std::endl;
    std::cout << "Output: " << output_file << std::endl;
    std::cout << "==================================================================" << std::endl;
    
    // Open output CSV
    std::ofstream csv_file(output_file);
    csv_file << "vertical_load,slip_angle,longitudinal_slip,camber_angle,velocity,"
             << "bekker_Kphi,bekker_Kc,bekker_n,mohr_cohesion,mohr_friction,janosi_shear,mesh_spacing,"
             << "Fz,Fx,Fy\n";
    std::mutex csv_mutex;
    
    // Setup terrain ranges
    TerrainRanges ranges;
    
    // Progress tracking
    std::atomic<int> completed(0);
    std::atomic<int> total_samples(0);
    auto start_time = std::chrono::steady_clock::now();
    
#ifdef CHRONO_OPENMP
    omp_set_num_threads(num_threads);
    
    #pragma omp parallel
    {
        // Each thread gets its own RNG
        int thread_id = omp_get_thread_num();
        std::mt19937 rng(seed + thread_id * 1000);
        
        #pragma omp for schedule(dynamic)
        for (int i = 0; i < num_sims; i++) {
            // Generate random terrain
            TerrainParams terrain = GenerateRandomTerrain(ranges, rng);
            
            // Run simulation
            auto measurements = RunVehicleSimulation(
                terrain, i, sim_duration, 3e-3, mesh_spacing, false
            );
            
            // Write results
            if (!measurements.empty()) {
                WriteMeasurementsToCSV(csv_file, csv_mutex, measurements);
                total_samples += measurements.size();
            }
            
            // Progress
            int done = ++completed;
            if (done % 10 == 0 || done == num_sims) {
                auto now = std::chrono::steady_clock::now();
                auto elapsed = std::chrono::duration<double>(now - start_time).count();
                double rate = done / elapsed;
                double remaining = (num_sims - done) / rate;
                
                #pragma omp critical
                {
                    std::cout << "\rProgress: " << done << "/" << num_sims 
                              << " (" << int(100.0 * done / num_sims) << "%)"
                              << " | Samples: " << total_samples.load()
                              << " | ETA: " << int(remaining) << "s" << std::flush;
                }
            }
        }
    }
#else
    // Serial fallback
    std::mt19937 rng(seed);
    
    for (int i = 0; i < num_sims; i++) {
        TerrainParams terrain = GenerateRandomTerrain(ranges, rng);
        auto measurements = RunVehicleSimulation(terrain, i, sim_duration, 3e-3, mesh_spacing, verbose);
        
        if (!measurements.empty()) {
            WriteMeasurementsToCSV(csv_file, csv_mutex, measurements);
            total_samples += measurements.size();
        }
        
        completed++;
        if (verbose && (i % 10 == 0)) {
            std::cout << "Progress: " << i << "/" << num_sims << std::endl;
        }
    }
#endif
    
    csv_file.close();
    
    auto end_time = std::chrono::steady_clock::now();
    auto total_time = std::chrono::duration<double>(end_time - start_time).count();
    
    std::cout << std::endl;
    std::cout << "==================================================================" << std::endl;
    std::cout << "Collection complete!" << std::endl;
    std::cout << "  Total simulations: " << num_sims << std::endl;
    std::cout << "  Total tire samples: " << total_samples.load() << std::endl;
    std::cout << "  Time: " << total_time << "s" << std::endl;
    std::cout << "  Rate: " << num_sims / total_time << " sims/s" << std::endl;
    std::cout << "  Output: " << output_file << std::endl;
    std::cout << "==================================================================" << std::endl;
    
    return 0;
}

File: project/SCM_Teleop/archive/collect_vehicle_data.py
#!/usr/bin/env python3
"""
Full Vehicle SCM Data Collection for NN Tire Model Training

Unlike the tire test rig, this collects data from actual vehicle simulation:
- Real tire loads from vehicle weight and load transfer
- Natural slip angles from cornering maneuvers
- Forces that match what the estimator sees

This should produce training data that better matches the estimator's environment.
"""

import pychrono as chrono
import pychrono.vehicle as veh
import numpy as np
import multiprocessing as mp
from multiprocessing import Pool, Value
import csv
import argparse
import time
import sys
from dataclasses import dataclass
from typing import List, Tuple


@dataclass
class TerrainParams:
    """Bekker soil parameters for SCM terrain"""
    bekker_Kphi: float
    bekker_Kc: float
    bekker_n: float
    mohr_cohesion: float
    mohr_friction: float  # degrees
    janosi_shear: float


@dataclass
class TireMeasurement:
    """Single tire measurement sample"""
    time: float
    axle: int  # 0=front, 1=rear
    side: int  # 0=left, 1=right
    slip_angle: float  # radians
    slip_ratio: float
    camber: float  # radians
    velocity: float  # m/s
    Fz: float  # vertical force
    Fx: float  # longitudinal force
    Fy: float  # lateral force
    # Terrain params
    bekker_Kphi: float
    bekker_Kc: float
    bekker_n: float
    mohr_cohesion: float
    mohr_friction: float
    janosi_shear: float


class TerrainRanges:
    """Parameter ranges for random terrain generation"""
    bekker_Kphi_min = 0.5e6
    bekker_Kphi_max = 4.0e6
    bekker_Kc_min = 0
    bekker_Kc_max = 20000
    bekker_n_min = 0.3
    bekker_n_max = 1.5
    mohr_cohesion_min = 0
    mohr_cohesion_max = 10000
    mohr_friction_min = 10  # degrees
    mohr_friction_max = 45  # degrees
    janosi_shear_min = 0.005
    janosi_shear_max = 0.05


def generate_random_terrain(rng: np.random.Generator) -> TerrainParams:
    """Generate random terrain parameters within ranges"""
    r = TerrainRanges
    return TerrainParams(
        bekker_Kphi=rng.uniform(r.bekker_Kphi_min, r.bekker_Kphi_max),
        bekker_Kc=rng.uniform(r.bekker_Kc_min, r.bekker_Kc_max),
        bekker_n=rng.uniform(r.bekker_n_min, r.bekker_n_max),
        mohr_cohesion=rng.uniform(r.mohr_cohesion_min, r.mohr_cohesion_max),
        mohr_friction=rng.uniform(r.mohr_friction_min, r.mohr_friction_max),
        janosi_shear=rng.uniform(r.janosi_shear_min, r.janosi_shear_max),
    )


def run_vehicle_simulation(
    terrain: TerrainParams,
    sample_idx: int,
    sim_duration: float = 15.0,
    step_size: float = 3e-3,
    mesh_spacing: float = 0.1,
    verbose: bool = False,
) -> List[TireMeasurement]:
    """Run a single vehicle simulation and collect tire data"""
    
    measurements = []
    
    # Set data paths
    chrono.SetChronoDataPath(chrono.GetChronoDataPath())
    veh.SetDataPath(chrono.GetChronoDataPath() + 'vehicle/')
    
    # Create HMMWV vehicle (it creates its own system internally)
    hmmwv = veh.HMMWV_Full()
    hmmwv.SetContactMethod(chrono.ChContactMethod_SMC)
    hmmwv.SetChassisFixed(False)
    hmmwv.SetInitPosition(chrono.ChCoordsysd(chrono.ChVector3d(0, 0, 0.6), chrono.QUNIT))
    hmmwv.SetEngineType(veh.EngineModelType_SHAFTS)
    hmmwv.SetTransmissionType(veh.TransmissionModelType_AUTOMATIC_SHAFTS)
    hmmwv.SetDriveType(veh.DrivelineTypeWV_AWD)
    hmmwv.SetTireType(veh.TireModelType_RIGID)  # RIGID for SCM interaction
    hmmwv.Initialize()
    
    # Get system from vehicle
    sys = hmmwv.GetSystem()
    sys.SetCollisionSystemType(chrono.ChCollisionSystem.Type_BULLET)
    sys.SetGravitationalAcceleration(chrono.ChVector3d(0, 0, -9.81))
    sys.SetSolverType(chrono.ChSolver.Type_BARZILAIBORWEIN)
    sys.GetSolver().AsIterative().SetMaxIterations(100)
    
    # Disable visualization 
    hmmwv.SetChassisVisualizationType(veh.VisualizationType_NONE)
    hmmwv.SetSuspensionVisualizationType(veh.VisualizationType_NONE)
    hmmwv.SetSteeringVisualizationType(veh.VisualizationType_NONE)
    hmmwv.SetWheelVisualizationType(veh.VisualizationType_NONE)
    hmmwv.SetTireVisualizationType(veh.VisualizationType_NONE)
    
    # Create SCM terrain
    scm_terrain = veh.SCMTerrain(sys)
    scm_terrain.SetSoilParameters(
        terrain.bekker_Kphi,
        terrain.bekker_Kc,
        terrain.bekker_n,
        terrain.mohr_cohesion,
        np.radians(terrain.mohr_friction),  # Convert to radians!
        terrain.janosi_shear,
        2e8,  # elastic stiffness
        3e4,  # damping
    )
    
    # Add moving patch for efficiency
    scm_terrain.AddMovingPatch(
        hmmwv.GetChassisBody(),
        chrono.ChVector3d(0, 0, 0),
        chrono.ChVector3d(6, 3, 1),
    )
    
    scm_terrain.SetPlotType(veh.SCMTerrain.PLOT_NONE, 0, 0)
    scm_terrain.Initialize(100.0, 30.0, mesh_spacing)
    
    # Vehicle parameters
    Lf = 1.689  # front axle to CG
    Lr = 1.689  # rear axle to CG
    
    # Simulation loop
    t = 0.0
    sample_count = 0
    sample_interval = 100  # Every 100 steps
    
    while t < sim_duration:
        # Generate driving inputs: acceleration then cornering
        if t < 2.0:
            steer = 0.0
            throttle = 0.5
        elif t < 5.0:
            steer = 0.3 * (t - 2.0) / 3.0
            throttle = 0.35
        elif t < 8.0:
            steer = 0.3
            throttle = 0.35
        elif t < 11.0:
            steer = 0.3 - 0.6 * (t - 8.0) / 3.0
            throttle = 0.35
        else:
            steer = -0.3
            throttle = 0.35
        
        # Apply inputs
        inputs = veh.DriverInputs()
        inputs.m_steering = steer
        inputs.m_throttle = throttle
        inputs.m_braking = 0.0
        
        # Synchronize and advance
        scm_terrain.Synchronize(t)
        hmmwv.Synchronize(t, inputs, scm_terrain)
        scm_terrain.Advance(step_size)
        hmmwv.Advance(step_size)
        
        # Sample tire data periodically after initial settling
        if t > 3.0 and (sample_count % sample_interval) == 0:
            # Get chassis state
            chassis = hmmwv.GetChassisBody()
            vel_global = chassis.GetPosDt()
            rot = chassis.GetRot()
            rot_inv = rot.GetInverse()
            vel_body = rot_inv.Rotate(vel_global)
            omega_vec = chassis.GetAngVelLocal()
            
            u = vel_body.x  # forward velocity
            v = vel_body.y  # lateral velocity
            omega = omega_vec.z  # yaw rate
            delta = steer * 0.5  # steering angle (approximate ratio)
            
            # Skip if vehicle too slow
            if u >= 2.0:
                # Get tire forces for all 4 wheels
                vehicle = hmmwv.GetVehicle()
                
                for axle in range(2):
                    for side in range(2):
                        side_enum = veh.LEFT if side == 0 else veh.RIGHT
                        tire = vehicle.GetTire(axle, side_enum)
                        
                        # Get tire force
                        tire_force = tire.ReportTireForce(scm_terrain)
                        force_local = rot_inv.Rotate(tire_force.force)
                        
                        # Compute slip angle
                        L = Lf if axle == 0 else -Lr
                        wheel_delta = delta if axle == 0 else 0.0
                        
                        eps = 0.5
                        u_safe = max(u, eps)
                        alpha = np.arctan2(v + L * omega, u_safe) - wheel_delta
                        
                        # Store measurement
                        m = TireMeasurement(
                            time=t,
                            axle=axle,
                            side=side,
                            slip_angle=alpha,
                            slip_ratio=0.0,
                            camber=0.0,
                            velocity=u,
                            Fz=force_local.z,
                            Fx=force_local.x,
                            Fy=force_local.y,
                            bekker_Kphi=terrain.bekker_Kphi,
                            bekker_Kc=terrain.bekker_Kc,
                            bekker_n=terrain.bekker_n,
                            mohr_cohesion=terrain.mohr_cohesion,
                            mohr_friction=terrain.mohr_friction,
                            janosi_shear=terrain.janosi_shear,
                        )
                        
                        # Filter for reasonable samples
                        if m.Fz > 1000 and abs(m.Fy) < 20000 and abs(alpha) > 0.01:
                            measurements.append(m)
        
        sample_count += 1
        t += step_size
    
    if verbose:
        print(f"Sample {sample_idx}: collected {len(measurements)} tire measurements "
              f"(n={terrain.bekker_n:.2f}, phi={terrain.mohr_friction:.1f}°)")
    
    return measurements


def run_single_sim_wrapper(args):
    """Wrapper for multiprocessing"""
    sample_idx, seed = args
    rng = np.random.default_rng(seed)
    terrain = generate_random_terrain(rng)
    
    try:
        measurements = run_vehicle_simulation(
            terrain=terrain,
            sample_idx=sample_idx,
            sim_duration=15.0,
            step_size=3e-3,
            mesh_spacing=0.1,
            verbose=False,
        )
        return measurements
    except Exception as e:
        print(f"Warning: Simulation {sample_idx} failed: {e}")
        return []


def main():
    parser = argparse.ArgumentParser(
        description="Collect tire force data from full vehicle SCM simulation"
    )
    parser.add_argument("-n", type=int, default=100, help="Number of simulations")
    parser.add_argument("-o", type=str, default="vehicle_tire_data.csv", help="Output CSV")
    parser.add_argument("-j", type=int, default=4, help="Parallel processes")
    parser.add_argument("-t", type=float, default=15.0, help="Simulation duration (s)")
    parser.add_argument("--seed", type=int, default=12345, help="Random seed")
    parser.add_argument("-v", action="store_true", help="Verbose output")
    args = parser.parse_args()
    
    print("=" * 66)
    print("Full Vehicle SCM Data Collection")
    print("=" * 66)
    print(f"Simulations: {args.n}")
    print(f"Processes: {args.j}")
    print(f"Duration each: {args.t}s")
    print(f"Output: {args.o}")
    print("=" * 66)
    
    # Open output file
    csv_file = open(args.o, "w", newline="")
    writer = csv.writer(csv_file)
    writer.writerow([
        "vertical_load", "slip_angle", "longitudinal_slip", "camber_angle", "velocity",
        "bekker_Kphi", "bekker_Kc", "bekker_n", "mohr_cohesion", "mohr_friction",
        "janosi_shear", "mesh_spacing", "Fz", "Fx", "Fy"
    ])
    
    start_time = time.time()
    all_measurements = []
    
    # Prepare simulation arguments
    sim_args = [(i, args.seed + i * 1000) for i in range(args.n)]
    
    if args.j > 1:
        # Multiprocessing
        with Pool(args.j) as pool:
            for i, measurements in enumerate(pool.imap_unordered(run_single_sim_wrapper, sim_args)):
                all_measurements.extend(measurements)
                
                # Write measurements to CSV
                for m in measurements:
                    writer.writerow([
                        m.Fz,  # vertical_load
                        m.slip_angle,
                        m.slip_ratio,
                        m.camber,
                        m.velocity,
                        m.bekker_Kphi,
                        m.bekker_Kc,
                        m.bekker_n,
                        m.mohr_cohesion,
                        m.mohr_friction,
                        m.janosi_shear,
                        0.1,  # mesh_spacing
                        m.Fz,
                        m.Fx,
                        m.Fy,
                    ])
                csv_file.flush()
                
                # Progress
                done = i + 1
                if done % 10 == 0 or done == args.n:
                    elapsed = time.time() - start_time
                    rate = done / elapsed
                    remaining = (args.n - done) / rate if rate > 0 else 0
                    print(f"\rProgress: {done}/{args.n} ({100*done//args.n}%) | "
                          f"Samples: {len(all_measurements)} | ETA: {int(remaining)}s", 
                          end="", flush=True)
    else:
        # Single process
        for i, (sample_idx, seed) in enumerate(sim_args):
            rng = np.random.default_rng(seed)
            terrain = generate_random_terrain(rng)
            measurements = run_vehicle_simulation(
                terrain=terrain,
                sample_idx=sample_idx,
                sim_duration=args.t,
                verbose=args.v,
            )
            all_measurements.extend(measurements)
            
            for m in measurements:
                writer.writerow([
                    m.Fz, m.slip_angle, m.slip_ratio, m.camber, m.velocity,
                    m.bekker_Kphi, m.bekker_Kc, m.bekker_n, m.mohr_cohesion,
                    m.mohr_friction, m.janosi_shear, 0.1, m.Fz, m.Fx, m.Fy,
                ])
            csv_file.flush()
            
            if args.v or (i + 1) % 10 == 0:
                print(f"Progress: {i+1}/{args.n}")
    
    csv_file.close()
    
    total_time = time.time() - start_time
    
    print()
    print("=" * 66)
    print("Collection complete!")
    print(f"  Total simulations: {args.n}")
    print(f"  Total tire samples: {len(all_measurements)}")
    print(f"  Time: {total_time:.1f}s")
    print(f"  Rate: {args.n/total_time:.2f} sims/s")
    print(f"  Output: {args.o}")
    print("=" * 66)


if __name__ == "__main__":
    main()
